### PR TITLE
Numeric arrays and collections attributes

### PR DESCRIPTION
**Pull request by YYChen01988**
## Goal

Custom attributes support array and collection with String, Int, Long, Double

## Changeset

Added custom attributes support types in array and collection in impl and open api
Collection support type any but not null

## Testing

Introducing new unit tests

### CODE REVIEW

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/AttributesJson.kt [hunk 1]
```diff
@@ -26,7 +26,56 @@ private fun toAttributeJson(value: Any, json: JsonWriter) {
...
+                        }
+                    }
+                }
+            }
+
+            is LongArray -> json.name("array").obj {
+                json.name("values").array {
+                    value.forEach { longValue ->
+                        json.obj {
+                            name("longValue").value(longValue.toString())
```
Comment (lemnik) [inline bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/AttributesJson.kt:64]:
```suggestion
                            name("intValue").value(longValue.toString())
```

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt [hunk 2]
```diff
@@ -45,4 +45,36 @@ public interface HasAttributes {
...
      */
     public fun setAttribute(name: String, value: Boolean)
+
+    /**
+     * Set a collection of values.
+     *
+     * @param name the attribute name
+     * @param value the value to set the attribute to
+     */
+    public fun setAttribute(name: String, value: Collection<Any>)
```
Comment (lemnik) [inline bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt:55]: The `Collection<Any>` should be nullable.

File: bugsnag-android-performance/src/test/java/com/bugsnag/android/performance/SpanJsonTest.kt [hunk 3]
```diff
@@ -71,6 +78,78 @@ class SpanJsonTest {
...
+                            }
+                        },
+                        {
+                            "key": "LongArray custom attributes", 
+                            "value": { 
+                                "array": {
+                                    "values": [
+                                        { "longValue": "11" },
+                                        { "longValue": "22" },
+                                        { "longValue": "33" }
```
Comment (lemnik) [inline bugsnag-android-performance/src/test/java/com/bugsnag/android/performance/SpanJsonTest.kt:101]:
```suggestion
                                        { "intValue": "11" },
                                        { "intValue": "22" },
                                        { "intValue": "33" }
```

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/JsonWriterExtensions.kt [hunk 4]
```diff
@@ -17,3 +17,40 @@ internal inline fun JsonWriter.array(builder: JsonWriter.() -> Unit): JsonWriter
 
     return this
 }
+
+internal fun writeIntArray(json: JsonWriter, value: IntArray) {
```
Comment (lemnik) [inline bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/JsonWriterExtensions.kt:21]: I think these methods are best kept in `AttributesJson.kt`

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt [hunk 5]
```diff
@@ -52,13 +52,37 @@ public interface HasAttributes {
...
 
     /**
-     * Set or clear a string collection attribute. Passing `null` will remove the attribute.
+     * Set a collection of values.
      *
      * @param name the attribute name
      * @param value the value to set the attribute to
      */
-    public fun setAttribute(name: String, value: Array<String>?)
+    public fun setAttribute(name: String, value: Collection<Any>?)
```
Comment (lemnik) [inline bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt:63]: I think it's worth documenting how these collections need to be structured. Something along the lines of:
> Set a collection of primitive values as an attribute. Every element should be of the same type, and only `String`, `Integer`, `Long`, `Double` and `Boolean` are supported. If the collection contains mixed values, `null`, or any other type: the behavior is undefined.

@tomlongridge thoughts?

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/JsonWriterExtensions.kt [hunk 6]
```diff
@@ -17,3 +17,5 @@ internal inline fun JsonWriter.array(builder: JsonWriter.() -> Unit): JsonWriter
 
     return this
 }
+
```
Comment (lemnik) [inline bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/JsonWriterExtensions.kt:20]: nit: I think we only need one newline at the end of the file

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/SpanImpl.kt [hunk 7]
```diff
@@ -176,13 +176,33 @@ public class SpanImpl internal constructor(
...
             attributes[name] = value
         }
     }
 
-    override fun setAttribute(name: String, value: Array<String>?) {
+    override fun setAttribute(name: String, value: Collection<Any>?) {
+        if (!isEnded()) {
+            if (value != null) {
+                attributes[name] = value
+            }
```
Comment (lemnik) [inline bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/internal/SpanImpl.kt:189]:
```suggestion
            attributes[name] = value
```

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt (hunk 5 above)
Comment (tomlongridge) [reply bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt:63]: Is it clear from the Otel spec that an array with some `null`s isn't allowed? I'd imagined a part-filled array would be acceptable and more common than actual mixed types.

Comment (lemnik) [reply bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt:63]: The [spec](https://opentelemetry.io/docs/specs/otel/common/) says:

> Attribute values of null are not valid and attempting to set a null value is undefined behavior.
> null values SHOULD NOT be allowed in arrays.

Comment (tomlongridge) [reply bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt:63]: Pretty clear then! Wording looks good to me.

File: bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt [hunk 8]
```diff
@@ -52,13 +52,41 @@ public interface HasAttributes {
...
-    public fun setAttribute(name: String, value: Collection<String>?)
+    public fun setAttribute(name: String, value: Array<String>?)
 
     /**
-     * Set or clear a string collection attribute. Passing `null` will remove the attribute.
+     * Set a collection of values.
+     * Set a collection of primitive values as an attribute.
+     * Every element should be of the same type, and only String, Integer, Long, Double and Boolean
+     * are supported.
+     * If the collection contains mixed values, null, or any other type: the behavior is undefined.
```
Comment (lemnik) [inline bugsnag-android-performance/src/main/kotlin/com/bugsnag/android/performance/HasAttributes.kt:62]:
```suggestion
     * Set a collection of primitive values as an attribute.
     * Every element should be of the same type, and only String, Integer, Long, Double and Boolean
     * are supported. If the collection contains mixed values, null, or any other type: the behavior is
     * undefined.
```

**Review by lemnik — APPROVED** [review]

**Pull request merged**